adal/src/main/java/com/microsoft/aad/adal/Logger.java

```diff
@@ -65,7 +54,31 @@ public static Logger getInstance() {
      * @param logLevel The {@link LogLevel} to be enabled for the diagnostic logging.
      */
     public void setLogLevel(final LogLevel logLevel) {
-        this.mLogLevel = logLevel;
+        switch (logLevel) {
```

Comment:
To lower the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of this function, this `switch` statement as well as the one used @ `L99` looks like a good use case for an [`EnumBiMap`](https://google.github.io/guava/releases/18.0/api/docs/com/google/common/collect/EnumBiMap.html) -- since we can't use this structure without adding the `Guava` dependency you _could_ cut down on the code here by initializing a `Map<Enum, Enum>` such that...

```java
Map<Enum, Enum> mLogLevelMap = new HashMap<>() {{
    put(LogLevel.Error, com.microsoft.identity.common.internal.logging.Logger.LogLevel.ERROR);
    put(com.microsoft.identity.common.internal.logging.Logger.LogLevel.ERROR, LogLevel.Error);
    // ... (do this for each LogLevel pairing)
}};

// Now the function @ L57 becomes
public void setLogLevel(final LogLevel logLevel) {
    // Avoid the switch block by calling get() on the Map
    com.microsoft.identity.common.internal.logging.Logger.getInstance()
        .setLogLevel(mLogLevelMap.get(logLevel));
}

// And the function @ L91 becomes
public synchronized void setExternalLogger(ILogger externalLogger) {
    com.microsoft.identity.common.internal.logging.Logger.getInstance().setExternalLogger(new ILoggerCallback() {
        @Override
        public void log(String tag, com.microsoft.identity.common.internal.logging.Logger.LogLevel logLevel, String message, boolean containsPII) {
            if (mExternalLogger != null) {
                if (!LoggerSettings.getAllowPii() && containsPII) {
                    return;
                } else {
                    // Avoid the switch block by calling get() on the Map
                    mExternalLogger.Log(tag, message, null, mLogLevelMap.get(logLevel), null);
                }
            }
        }
    });

    mExternalLogger = externalLogger;
}
```

(As a word of caution, I haven't tested this code)

Reply:
Sorry, I took it back. I reconsidered this implementation. Actually the mapping from adal loglevel to common-core loglevel is only used once in setLogLevel. On the other hand, the mapping from common-core loglevel to adal loglevel also only happen once in setExternalLogger. I do not think it would be a good idea to hold the mixed mapping between common-core and adal.